### PR TITLE
fix(store): do not recompute unmounted atoms eagerly

### PR DESCRIPTION
**Pull request by dai-shi**
Fix #2789

Another take for #2845

### CODE REVIEW

File: tests/vanilla/memoryleaks.test.ts [hunk 1]
```diff
@@ -91,3 +91,45 @@ describe('test memory leaks (with subscribe)', () => {
...
+    const store = createStore()
+    const detector = new LeakDetector(store.get(a).obj)
+    let unsub: (() => void) | undefined = store.sub(c, () => {}) // mounts c,a
+    store.set(w)
+    unsub()
+    unsub = undefined
+    a = undefined
+    expect(await detector.isLeaking()).toBe(false)
+  })
+})
```
Comment (dai-shi) [inline tests/vanilla/memoryleaks.test.ts:135]: @dmaskasky Does your branch #2845 pass these tests, right?

Comment (dmaskasky) [reply tests/vanilla/memoryleaks.test.ts:135]: Yes, all tests pass.

File: src/vanilla/store.ts [hunk 2]
```diff
@@ -418,14 +418,17 @@ const buildStore = (
...
     pending: Pending,
     atom: Atom<Value>,
     atomState: AtomState<Value>,
   ): Map<AnyAtom, AtomState> => {
     const dependents = new Map<AnyAtom, AtomState>()
     for (const a of atomState.m?.t || []) {
-      dependents.set(a, getAtomState(a))
+      const aState = getAtomState(a)
+      if (aState.m) {
+        dependents.set(a, aState)
```
Comment (dmaskasky) [inline src/vanilla/store.ts:430]: Could you clarify this section? What specific behavior are you aiming to achieve here?

File: tests/vanilla/dependency.test.tsx [hunk 3]
```diff
@@ -346,3 +346,62 @@ it('can read sync derived atom in write without initializing', () => {
...
+    return 'a'
+  })
+  const w = atom(null, (get) => get(a))
+  const store = createStore()
+  store.sub(a, () => {}) // mounts a
+  store.set(w)
+  expect(aReadCount).toBe(1)
+  store.set(w)
+  expect(aReadCount).toBe(1)
+})
```
Comment (dai-shi) [inline tests/vanilla/dependency.test.tsx:407]: Does #2845 pass these too?

**Comment by dai-shi** [comment]
I guess the two tests

- "can cache reading an atom in write function"
- "atom only read in write function"

are conflicting. So, it's a trade-off and our design choice.

File: tests/vanilla/dependency.test.tsx (hunk 3 above)
Comment (dmaskasky) [reply tests/vanilla/dependency.test.tsx:407]: Yes, #2845 passes all tests.

Comment (dai-shi) [reply tests/vanilla/dependency.test.tsx:407]: Let me cherry-pick 753af9fbd8af0704955c424259bd48c3a353aa63 in #2845.

**Pull request merged**